Source/EasyNetQ/DefaultMessageSerializationStrategy.cs

```diff
@@ -44,9 +44,13 @@ public SerializedMessage SerializeMessage(IMessage message)
     }
 
     /// <inheritdoc />
-    public IMessage DeserializeMessage(MessageProperties properties, in ReadOnlyMemory<byte> body)
+    public IMessage DeserializeMessage(
+        MessageProperties properties, in ReadOnlyMemory<byte> body, Type? fallbackMessageType = null
+    )
     {
-        var messageType = typeNameSerializer.Deserialize(properties.Type!);
```

Comment:
If I understand correctly there was exception before since `properties.Type` was `null`. Right?

Reply:
Right.

Reply:
And the core problem is some other platform/library does not set `type` property from publishing side to derive .NET type on consuming side.

Reply:
Correct.

And actually it is a redundant requirements for the simple case when there is a single queue and a single handler(no need to dispatch based on type).

Reply:
If so I think it would be better to incapsulate such behavior in a new serialization strategy class rather than cluttering public API with awkward `fallbackMessageType `. `fallbackMessageType ` is implementation detail.

Reply:
> If so I think it would be better to incapsulate such behavior in a new serialization strategy class rather than cluttering public API with awkward `fallbackMessageType `. `fallbackMessageType ` is implementation detail.

I have the same feelings about the PR. Do you have a more detailed view on how this brand new strategy should look like? How it will be aware about a type which a handler expect to have?

Reply:
I think that what you really want is to somehow pass **exactly** `T` (not first, not last) generic argument down to serialization strategy instance.

> it is a redundant requirements for the simple case when there is a single queue and a single handler(no need to dispatch based on type).

Then all the more, will it be the best decision to just register custom serialization strategy/typenameserializer and just return hardcoded type? It looks strange that the core library begins to break its interfaces in order to solve a special edge-case.

Reply:
> How it will be aware about a type which a handler expect to have?

Just hardcode it and return as is.

Reply:
In essence, the consuming party knows this type. It is fixed. As you say, this is a simple case of interaction.

Reply:
Also note that your hack work only for overloads of `Consume` with generic parameter.

Reply:
I agree with you mostly, but it doesn't help me to understand a solution you are talking about.

Please, submit a PR to illustrate your idea.

PS Probably I forgot to mention that I am not inclined to merge this PR as it is.

Reply:
```
public FallbackMessageSerializationStrategy : DefaultMessageSerializationStrategy
{
 public FallbackMessageSerializationStrategy(
        ITypeNameSerializer typeNameSerializer,
        ISerializer serializer,
        ICorrelationIdGenerationStrategy correlationIdGenerator,
        Type fallbackType
    ) : base(typeNameSerializer, serializer, correlationIdGenerator)
{
//store fallbackType
}

   public override IMessage DeserializeMessage(MessageProperties properties, in ReadOnlyMemory<byte> body)
    {
         if (properties.Type is null)
              properties.Type = typeNameSerializer.Serialize(fallbackType);
        return base.DeserializeMessage(properties, body);
    }
}
```

if you want to preserve behavior of dispatching the value of `type` property when it was set. Plus you may want to provide some fluent API to setup `FallbackMessageSerializationStrategy ` like `.EnableNewtonsoftJson()` or `.EnableAlwaysNackWithoutRequeueConsumerErrorStrategy()`.

No need to break core interfaces.

Reply:
A fallback type could be different per a consume call.

Reply:
In the general case, fallbackType depends on the queue from which messages are consumed, so it will not be superfluous to think about how to provide access to this information at the lower API levels.

Reply:
> A fallback type could be different per a consume call.

Of course. We write comments in parallel. But again - you said
> redundant requirements for the simple case when there is a **single queue** and a **single handler**(no need to dispatch based on type).

Reply:
I meant not a constant global fallback type, but a fallback type when a single handler is registered to consumer from a queue.